### PR TITLE
api: add Client / Transport for clientID-authentication

### PR DESCRIPTION
**Pull request by rjeczalik**
~~Depends on: https://github.com/koding/koding/pull/9786~~

### CODE REVIEW

File: go/src/koding/kites/kloud/kloud/kloud.go [hunk 1]
```diff
@@ -235,12 +237,46 @@ func New(conf *Config) (*Kloud, error) {
...
+		return &socialapi.Session{
+			Username: s.Username,
+			Team:     s.GroupName,
+			ClientID: s.ClientId,
+		}, nil
+	}
+
+	transport := &socialapi.Transport{
+		RoundTripper: storeOpts.Client.Transport,
+		Host:         remoteURL.Host,
```
Comment (ppknap) [inline go/src/koding/kites/kloud/kloud/kloud.go]: this will `panic` if remoteURL is nil (you're checking that case in line 267)

File: go/src/koding/socialapi/transport.go [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+// requests with the given session.
+//
+// Session is expected to be have a valid clientID. In order to be
+// able to refresh the session if it gets invalidated, the session
+// must also contain Username / Team information.
+//
+// This is usefull for api clients that give no control over HTTP requests,
+// like the swagger-generated ones (remote.api).
+func (t *Transport) NewSingleClient(session *Session) http.RoundTripper {
+	return &DirectorTransport{
```
Comment (ppknap) [inline go/src/koding/socialapi/transport.go]: Hmm don't know the current status but this may interest you(first 40secs): https://youtu.be/FARQMJndUn0?t=1151

Comment (rjeczalik) [inline go/src/koding/socialapi/transport.go]: The point is no-one should type-assert on a concrete type, rather do interface upgrade instead.

File: go/src/koding/kites/kloud/kloud/kloud.go (hunk 1 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/kloud/kloud.go]: Done, thanks!

File: go/src/koding/socialapi/transport.go (hunk 2 above)
Comment (ppknap) [inline go/src/koding/socialapi/transport.go]: This is only fyi that DefaultTransport cannot be replaced. Nothing here to fix.

File: go/src/koding/socialapi/cache_test.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	var auth = NewFakeAuth()
+
+	users := []*socialapi.Session{
+		{Username: "user1", Team: "foobar"},
+		{Username: "user2", Team: "foobar"},
+		{Username: "user3", Team: "foobar"},
+		{Username: "user", Team: "team"},
+	}
+
+	cases := []struct {
```
Comment (ppknap) [inline go/src/koding/socialapi/cache_test.go]: could you change it to `map[string]struct { .... }` ?

File: go/src/koding/socialapi/transport.go [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const maxBodyLen = 2 << 20 // 2 MiB
```
Comment (ppknap) [inline go/src/koding/socialapi/transport.go]: :+1:

File: go/src/koding/socialapi/transport.go [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+	Team string
+}
+
+// Key is used for SessionCache to cache sessions by keys.
+func (s *Session) Key() string {
+	return s.Team + "/" + s.Username
+}
+
+// Valid implements the stack.Validator interface.
+func (s *Session) Valid() error {
```
Comment (ppknap) [inline go/src/koding/socialapi/transport.go]: do we allow empty teams and usernames? If not, could you add them to validator too?

**Review by ppknap — APPROVED** [review]

File: go/src/koding/socialapi/transport.go (hunk 5 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/transport.go]: Yes, we currently do.

File: go/src/koding/kites/kloud/kloud/kloud.go [hunk 6]
```diff
@@ -30,6 +31,7 @@ import (
 	"koding/kites/kloud/terraformer"
 	"koding/kites/kloud/userdata"
 	"koding/remoteapi"
+	"koding/socialapi"
```
Comment (cihangir) [inline go/src/koding/kites/kloud/kloud/kloud.go]: can we rename socialapi to api or clients? socialapi is one of the service name and we have the same package path?

File: go/src/koding/kites/kloud/kloud/kloud.go [hunk 7]
```diff
@@ -235,12 +237,45 @@ func New(conf *Config) (*Kloud, error) {
...
+	if remoteURL != nil {
+		transport.Host = remoteURL.Host
+
+		localRemoteURL := *remoteURL
+
+		// TODO(rjeczalik): Add PrivateRemoteURL to koding/config.
+		if _, port, err := net.SplitHostPort(localRemoteURL.Host); err == nil {
+			localRemoteURL.Host = net.JoinHostPort("127.0.0.1", port)
+		} else {
+			localRemoteURL.Host = "127.0.0.1"
```
Comment (cihangir) [inline go/src/koding/kites/kloud/kloud/kloud.go]: lets remove the hardcoded configurations and read them from config

File: go/src/koding/socialapi/cache.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+	// new session with Refresh set to true.
+	AuthFunc AuthFunc
+
+	// Storage is an external storage for storing session.
+	//
+	// If nil, in-memory map is going to be used instead.
+	Storage Storage
+
+	cacheMu sync.RWMutex
+	cache   map[string]*Session
```
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: if we have Storage what is the use of cache here? lets use kodig/cache whenever possible

File: go/src/koding/socialapi/cache.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+//
+//   cache := socialapi.NewSessionCache(authFn)
+//
+//   t := &socialapi.Transport{
+//       AuthFunc: cache.Auth,
+//   }
+//
+func (s *SessionCache) Auth(opts *AuthOptions) (*Session, error) {
+	if s.AuthFunc == nil {
+		panic("socialapi: AuthFunc is nil")
```
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: this is not socialapi :)

File: go/src/koding/socialapi/cache.go (hunk 8 above)
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: you can implement MultiCache which fallbacks to another layer if required

File: go/src/koding/socialapi/cache.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+	session := opts.Session
+
+	// Early return - return existing session if it's valid
+	// and we were not ask to invalidate it.
+	if err := session.Valid(); err == nil && !opts.Refresh {
+		return session, nil
+	}
+
+	if session == nil {
```
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: better to check session before validating it

File: go/src/koding/socialapi/cache.go [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+		return nil, errors.New("cannot determine user session")
+	}
+
+	if !opts.Refresh {
```
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: could you seperate refreshing logic into another function?

File: go/src/koding/socialapi/cache.go [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+	//   for the next request
+	//
+	// System-wise Kloud uses in-memory cache while KD and Klient
+	// use BoltDB-backed cache. Since the latter two take exclusive
+	// lock of the database file, the only reason storage can
+	// fail are filesystem errors - there is no recovery from
+	// that, and falling back to memory makes sense only for
+	// Klient.
+	s.cacheMu.Lock()
+	if opts.Refresh {
```
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: you already checked if this is a refresh operation on line 87

File: go/src/koding/socialapi/cache.go [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	if err == nil {
+		_ = s.set(session)
+	}
+	s.cacheMu.Unlock()
+
+	return session, err
+}
+
+func (s *SessionCache) get(session *Session) error {
```
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: "get" is a misleading word for this function, what about something like populate/ensure?

File: go/src/koding/socialapi/cache.go [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+// when requested session was not found.
+var ErrSessionNotFound = errors.New("session not found")
+
+// Storage is an interface for external session storage.
+type Storage interface {
+	// Get gets a session.
+	//
+	// It returns ErrSessionNotFound if requested
+	// session was not cached.
+	Get(*Session) error
```
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: Get should return the session

File: go/src/koding/socialapi/helper_test.go [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+	for _, s := range fa.Sessions {
+		if s.ClientID == req.ClientID {
+			session = s
+			break
+		}
+	}
+	fa.mu.RUnlock()
+
+	p, err := ioutil.ReadAll(r.Body)
+	log.Printf("FakeAuth.GetSession: read body: p=%q, err=%v", p, err)
```
Comment (cihangir) [inline go/src/koding/socialapi/helper_test.go]: could you use our logger or remove this?

File: go/src/koding/socialapi/helper_test.go [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+
+	"koding/kites/kloud/utils"
+	"koding/socialapi"
+)
+
+type FakeAuth struct {
```
Comment (cihangir) [inline go/src/koding/socialapi/helper_test.go]: instead of mocking could we use the real code to test?

File: go/src/koding/socialapi/helper_test.go [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+type contextKey struct {
+	name string
+}
+
+var (
+	ErrorContextKey    = &contextKey{"fake-error"}
+	ResponseContextKey = &contextKey{"fake-response"}
+)
+
+type FakeTransport struct {
```
Comment (cihangir) [inline go/src/koding/socialapi/helper_test.go]: instead of the fake transport lets use the real one

File: go/src/koding/socialapi/presence/presence.go [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"errors"
+	"net/http"
+	"net/url"
+
+	"koding/socialapi"
+)
+
+type Client struct {
```
Comment (cihangir) [inline go/src/koding/socialapi/presence/presence.go]: could you use the provided client in the socialapi/workers/presence/client here https://github.com/koding/koding/pull/9863/files#diff-9da7fe549db0224bc2de4a603ce84d51R47 (after you merge :D )

File: go/src/koding/socialapi/transport_test.go [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+		},
+	}}
+
+	rec := &AuthRecorder{
+		AuthFunc: cache.Auth,
+	}
+
+	client := http.Client{
+		Transport: &socialapi.Transport{
+			RoundTripper: &FakeTransport{},
```
Comment (cihangir) [inline go/src/koding/socialapi/transport_test.go]: could you use the real transport here?

File: go/src/koding/socialapi/transport.go [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+	return &DirectorTransport{
+		RoundTripper: t,
+		Director: func(req *http.Request) {
+			*req = *session.WithRequest(req)
+		},
+	}
+}
+
+// RoundTrip implements the http.RoundTripper interface.
+func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (cihangir) [inline go/src/koding/socialapi/transport.go]: not checking this function for now since may need to refactor to be able to use in your tests

File: go/src/koding/socialapi/transport_test.go [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+	"net/http"
+	"strings"
+	"testing"
+
+	"koding/kites/tunnelproxy/discover/discovertest"
+	"koding/socialapi"
+)
+
+func TestTransport(t *testing.T) {
+	var storage TrxStorage
```
Comment (cihangir) [inline go/src/koding/socialapi/transport_test.go]: could you use the real storage here?

**Review by cihangir — CHANGES_REQUESTED** [review]
instead of mocking, lets use the the real functionality in the tests

File: go/src/koding/socialapi/transport_test.go (hunk 19 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/transport_test.go]: No, I can't. The real transport is wrapped with FakeTransport. FakeTransport is used to fake transport errors.

File: go/src/koding/socialapi/presence/presence.go (hunk 18 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/presence/presence.go]: Not without refactoring it. I need to have access to *http.Request. I can refactor it thought to allow for that. For now the requests are internally created, I order to request authentication I'd need to have access to requests' context:

https://github.com/koding/koding/pull/9863/files#diff-9da7fe549db0224bc2de4a603ce84d51R63

File: go/src/koding/socialapi/transport_test.go (hunk 21 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/transport_test.go]: No, I can't. The TrxStorage is kind of recording storage. It records all calls to Set / Get / Delete functions and then we verify in tests.

File: go/src/koding/socialapi/transport_test.go (hunk 19 above)
Comment (cihangir) [inline go/src/koding/socialapi/transport_test.go]: i believe we can use after a bit of refactoring. 

As we talked previously and all agreed on, instead of mocking lets use the real functionality in our tests.

File: go/src/koding/socialapi/transport.go (hunk 20 above)
Comment (cihangir) [inline go/src/koding/socialapi/transport.go]: btw i have alreday checked but not commenting

File: go/src/koding/socialapi/cache.go (hunk 8 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: > lets use kodig/cache whenever possible

The `cache` is default memory cache if Storage is nil. I avoided koding/cache (the `Memory` one) since I did not like it implementation [0] - moreover the implementation is so trivial (map + mutex) that I thought it does not varrant pulling in external dependency. 

[0] should have sync.RWMutex instead - https://github.com/koding/cache/blob/master/memory.go

File: go/src/koding/socialapi/cache.go (hunk 12 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: Yes, and here I check it again under critical section in order to update storage.

File: go/src/koding/socialapi/helper_test.go (hunk 16 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/helper_test.go]: Which real code?

File: go/src/koding/socialapi/transport_test.go (hunk 19 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/transport_test.go]: How do you want to test for transport errors using real functionality?

File: go/src/koding/kites/kloud/kloud/kloud.go (hunk 6 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/kloud/kloud.go]: Done in https://github.com/koding/koding/pull/9870/commits/4ff0b9494a13d1dfaf3680748d4209c044bc6646.

Renamed to`api`. Note for future - we're going to merge the following packages:

1) `koding/klientctl/endpoint`
2) `koding/remoteapi`

Under `koding/api` in future. For now we're not done with 1) API-wise so we do not see yet how would the final result look like. Just fyi.

File: go/src/koding/kites/kloud/kloud/kloud.go (hunk 7 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/kloud/kloud.go]: Done in https://github.com/koding/koding/pull/9870/commits/5e4f7a6c88f28b9bea1222e204ffeb719e410ebb.

File: go/src/koding/socialapi/helper_test.go (hunk 15 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/helper_test.go]: Done in 3b9b8a9bc2dfbe425b675c07bfb75f67d42e5241.

File: go/src/koding/socialapi/cache.go (hunk 13 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: Done - `get` is removed by a1ff5b919a8d5a2c32d7b6ceb98a666297ce7677.

File: go/src/koding/socialapi/cache.go (hunk 10 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: Done in dd3d5b9edacdcd39fc8d5a2bc7f1209192869ce4.

File: go/src/koding/socialapi/cache.go (hunk 9 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: Done in 6305645866dbda08d850ca8fcb5b24a07d86963d.

File: go/src/koding/socialapi/cache.go (hunk 14 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: Done in 0d10ce61f121760fcd4ce0a5b02441b8961967a5.

**Comment by rjeczalik** [comment]
@cihangir Addressed all except comments about FakeAuth and FakeTransport.

- there's no real replacement for FakeAuth - in production we use 1a) `auth.login` kite method or 1b) `modelhelper.FetchOrCreateSession` - using 1a) is not possible as there's no mean to start koding as a part of this unittest; using 1b) would not change much in my opinion, as the only thing that changes is the storage - mongo instead of memory, and `api` package responsibility is not to test mongo

- if we want to get rid of FakeTransport, we also get rid of [this](https://github.com/koding/koding/pull/9851/files#diff-c519253ffa5abcd1f4885d8a72242b96R66) and [this](https://github.com/koding/koding/pull/9851/files#diff-c519253ffa5abcd1f4885d8a72242b96R74) test case.

How to proceed?

**Comment by ppknap** [comment]
@rjeczalik I wrote a simple [test case](https://gist.github.com/ppknap/6e2974b7999901be596953e51b09d901) for cache type. Is it valid? If yes, this means that the cache itself or test fixtures have [data-races](https://gist.github.com/ppknap/b708b1af302238d5132888affd630a52).

**Comment by rjeczalik** [comment]
> If yes, this means that the cache itself or test fixtures have data-races

@ppknap Thanks.

I've included your test and a fix in https://github.com/koding/koding/pull/9870/commits/8631e946e5a686b401576558c6154ac6b8e84380.

File: go/src/koding/kites/kloud/kloud/kloud.go (hunk 7 above)
Comment (cihangir) [inline go/src/koding/kites/kloud/kloud/kloud.go]: thanks

File: go/src/koding/socialapi/cache.go (hunk 8 above)
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: i believe you wrote these in frustration :) the Rafal i know does not complain instead fixes them whenever he see a problem. 

Other than that i would strongly suggest you to use LRU instead of simple memory cache, since the talk all started from leaking memory not the trivial implementation itself. In the previous-previous talk we both were OK with a simple map

Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: > Other than that i would strongly suggest you to use LRU instead of simple memory cache, since the talk all started from leaking memory not the trivial implementation itself.

Sure, but let's do this in a separate PR, can we?

File: go/src/koding/socialapi/cache.go (hunk 12 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: The method was reworked - is there still anything do to here?

File: go/src/koding/socialapi/helper_test.go (hunk 16 above)
Comment (cihangir) [inline go/src/koding/socialapi/helper_test.go]: socialapi.NewCache

File: go/src/koding/socialapi/presence/presence.go (hunk 18 above)
Comment (cihangir) [inline go/src/koding/socialapi/presence/presence.go]: why do you need to refactor? this package has the exact same functionality. Why do you need to access to http.Request?

File: go/src/koding/socialapi/transport_test.go (hunk 21 above)
Comment (cihangir) [inline go/src/koding/socialapi/transport_test.go]: i believe you can. it is same implementation with an inmemory storage with an GetAll(maybe?) function, and it would eliminate some of the code

File: go/src/koding/socialapi/presence/presence.go (hunk 18 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/presence/presence.go]: To add api.SessionContextKey to http.Request.Context here:

https://github.com/koding/koding/pull/9851/files#diff-64803beca13535937999dd1daec4743cR22

This way single transport knows for which user it should get the ClientID.

File: go/src/koding/socialapi/transport_test.go (hunk 21 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/transport_test.go]: Done in https://github.com/koding/koding/commit/28fc46282c43fa1c87fbec1908bd19894d652cdc.

File: go/src/koding/socialapi/cache.go (hunk 8 above)
Comment (cihangir) [inline go/src/koding/socialapi/cache.go]: If we are going to start using this right after we merge, lets implement, otherwise we can do later

File: go/src/koding/socialapi/transport_test.go (hunk 19 above)
Comment (cihangir) [inline go/src/koding/socialapi/transport_test.go]: i was thinking smth similar to these https://github.com/golang/go/blob/master/src/net/http/transport.go#L1811-L1821

Comment (cihangir) [inline go/src/koding/socialapi/transport_test.go]: I have checked the code test coverage, it is in good shape, it can stay as is. 
on the other i couldnt see the integration of these into wercker, did i miss it or not done yet?

File: go/src/koding/socialapi/presence/presence.go (hunk 18 above)
Comment (cihangir) [inline go/src/koding/socialapi/presence/presence.go]: ok i see the problem, you dont have access to `clientId` since it is in the transport cache/auther.

on the other hand NewRequest is exported https://github.com/koding/koding/pull/9863/files#diff-9da7fe549db0224bc2de4a603ce84d51R72 so you have acccess to the http.request itself

File: go/src/koding/socialapi/transport_test.go (hunk 19 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/transport_test.go]: Done in https://github.com/koding/koding/commit/bb29400bcb4a23c9c2b1a863372ec4208d8d636d.

File: go/src/koding/socialapi/cache.go (hunk 8 above)
Comment (rjeczalik) [inline go/src/koding/socialapi/cache.go]: Done in https://github.com/koding/koding/commit/6d0d33f7d7fab14a8bf0ee274df6e8ab78f12344.

**Comment by rjeczalik** [comment]
@cihangir all addressed, ptal

File: go/src/koding/socialapi/transport_test.go (hunk 19 above)
Comment (cihangir) [inline go/src/koding/socialapi/transport_test.go]: > they are not faking or mocking any functionality.

which does not mean you cant, you can think them as small adapters and do what you want to do

for the hook debate (since you said it makes it worse), i find easier to reason about the cases/logic since they are in the code itself, not separated into another layer. i dont support having many hooks in the code itself too but i am also against having a big function and to be able to test adding another level of complexity with trying to fake the logical parts of it, which incidentally may lead to bugs in the tests itself(another question comes as how/should we (are going to) test the tests?). I would probably go with unit tests and smaller functions instead of having bigger components and mocks. 

in the end everyone has their own taste and there is no universal truth, at least i know of.

**Review by cihangir — APPROVED** [review]

**Comment by cihangir** [comment]
Thanks @rjeczalik

**Comment by cihangir** [comment]
for the followers of this thread, we decided not to have big PRs from now on, which causes lengthy discussions and making things harder to reason about while reviewing, following up with the comments and their resolutions.

File: go/src/koding/socialapi/presence/presence.go (hunk 18 above)
Comment (cihangir) [inline go/src/koding/socialapi/presence/presence.go]: > I don't see a clean way to use the client.

yep i stated that as a problem in my previous comment
